### PR TITLE
fix(pty): recover sessions after socket send failures

### LINKED ISSUE #110849: PtySession.attach() commits attached state before the replay write — a client dropping during attach leaves the PTY permanently unreapable
### Summary

`PtySession.attach()` commits the attached state *before* it writes the replay buffer to the socket, and the caller (`pty_ws`) performs that call **outside** any `try/finally`. A client that drops during the attach replay therefore raises out of the handler with the session already marked `attached=True, last_detached_at=None` and never reaches `PTY_REGISTRY.detach()`.

That state is the one state `reap_idle()` can never reclaim:

* `not s.alive` is `False` — the PTY child is perfectly alive;
* the second clause requires `not s.attached` **and** `s.last_detached_at is not None` — both are false.

`_reap_one_idle_or_raise()` has the same `not attached` requirement, and the 30-minute TTL is anchored to `last_detached_at`, which stays `None` forever. The PTY process group (`node ui-tui/dist/entry.js` + its `python -m tui_gateway.entry` child) then lives indefinitely, holding a tty, the PTY master fd, its loopback sockets and the session's work.

This is an independent entry into "attached forever" — it needs no reverse proxy, no disabled WS ping, and no SIGHUP cascade; the client only has to disappear between `attach` bookkeeping and the first byte of the replay.

### Where (verified on `main` @ a89c1e11)

`hermes_cli/pty_session.py` (attach, ~L98-L114):

```python
if self._ws is not ws:
    await _close_ws(self._ws, WS_CLOSE_SUPERSEDED)
self._ws = ws
self._attach_generation += 1
self.attached = True
self.last_detached_at = None
if snap := self.buffer.snapshot():
    await ws.send_bytes(snap)          # <- raises here, state already committed
```

`hermes_cli/web_routers/chat_ws.py` (L516) — the call site is outside the `try` that owns the detach; `PTY_REGISTRY.detach()` only runs in the writer loop's `finally` further down:

```python
if not await session.attach(ws, force_redraw=not _created):
    await _close_stalled_pty_input(ws, path="keepalive-redraw")
    PTY_REGISTRY.detach(attach_token, ws)
    return
```

### Evidence

Dashboard server log (3 occurrences, same shape), client-side `/api/pty` socket:

```
chat_ws.py:516 in pty_ws
pty_session.py:111 in attach
    await ws.send_bytes(snap)
starlette/websockets.py:89 in send -> raise WebSocketDisconnect(code=1006)
...
uvicorn.protocols.utils.ClientDisconnected
```

Field accumulation on one `hermes dashboard` with a single real terminal (`who` → console + one SSH session):

* peak snapshot: **18** `node ui-tui/dist/entry.js` and **15** `python -m tui_gateway.entry` processes, **15** ttys held, **15** PTY master fds in the dashboard process;
* all ESTABLISHED sockets on the dashboard port were held by the dashboard process itself (no browser clients at all) — i.e. no live viewer anywhere, yet every session stayed attached;
* one orphaned `ui-tui` node (no `tui_gateway` child) survived from the same day five days earlier and was still present in every later snapshot;
* a rescan 45 minutes later showed the total fluctuating (8 nodes / 7 children) but the oldest orphan unchanged — the leak is cumulative, not a one-off.

### Impact

Unbounded process / PTY / fd leak per dropped client, plus indirectly a stuck session lease: a stranded `tui_gateway` with an unfinished turn keeps refreshing its `session_turn_leases` row, which is invisible to watchdog checks keyed on "lease expired" or "owner pid dead" (the expiry is fresh and the pid is alive).

### Proposed fix (any of)

1. In `attach()`, do not commit `_ws`/`attached`/`last_detached_at` until the replay buffer has been delivered — send first, then record the attachment (or roll back in `except`).
2. In `pty_ws`, wrap the `session.attach(...)` call in `try/except` and `PTY_REGISTRY.detach(attach_token, ws)` before re-raising (or fold the attach into the writer loop's `try/finally`).
3. In `reap_idle()`, add a criterion independent of `attached` (e.g. an attach timestamp / socket-liveness probe) so a session whose viewer is gone is eventually reclaimed even if the detach was never observed.

Regression test: a `FakeWS` whose `send_bytes` raises → after the failed attach the session must still be reclaimable (`attached is False` or `last_detached_at is not None`). The existing suite pins the opposite invariant (`tests/test_pty_session.py` — "attached → not reapable") but does not cover an aborted attach.

### Related

* #110164 — orphaned `ui-tui` nodes with dead gateway children; the attach-abort path above is another, proxy-independent source of the same accumulation.
* #96418 — attached session never reaped when the loopback bind disables the WS ping (external trigger); this report is the in-process trigger and does not require the ping to be off.
* #106400 — `reap_idle()` `pop(key)` without default; fixed on `main` (a89c1e11) and **not** the same path as this report.
* PR #110168 — keeps the embedded dashboard TUI alive on SIGHUP; unrelated to the stale-attached state.

### PR DESCRIPTION
**Pull request by MohamadKanso**
Fixes #110849

What changed
- mark a session detached when the initial replay socket send fails
- mark it detached when a live PTY output send fails
- keep the existing reaper responsible for cleaning up the session
- add regression tests for both cases

Why
A dropped client could leave a PTY marked attached forever. The normal reaper could not reclaim it, so the PTY process and file descriptors could stay around indefinitely.

Tests
- ./scripts/run_tests.sh tests/hermes_cli/test_pty_session.py
- ruff check hermes_cli/pty_session.py tests/hermes_cli/test_pty_session.py
- git diff --check

### CODE REVIEW

**Comment by alt-glitch** [comment]
> *This was generated by AI during triage.*

Related: #110861 is a competing fix for #110849 (attach-replay rollback only). This PR additionally handles `_drain` send failures. Maintainers should pick one.

**Comment by kyssta-exe** [comment]
## Review: recover sessions after socket send failures

**Summary:** Small, focused fix for #110849: both PTY send sites (`attach()` replay and `_drain()` live forwarding) now mark the session detached when the socket write fails, so the existing TTL reaper can reclaim sessions a dropped client would otherwise pin forever. Two regression tests included.

**What changed**
- `hermes_cli/pty_session.py`: `_drain()` snapshots `ws = self._ws` before send and calls `self.detach(ws)` on failure (was silent `pass`); `attach()` wraps replay `send_bytes` in try/except that detaches then re-raises.
- `tests/hermes_cli/test_pty_session.py`: `FailingWS` harness + replay-failure and live-send-failure detach tests.

**Strengths**
- Reuses the identity-checked `detach(ws)`, so a stale socket's failure can't detach a replacement viewer — the key race is handled.
- Capturing `ws` before the await in `_drain` avoids acting on a socket swapped mid-send.
- Minimal diff, clear why/comment trail back to the issue.

**Findings**
- **Coordination needed:** #110861 and #110863 fix the same issue (#110849) on the same file with different trade-offs (return-False vs re-raise; replay-only vs generation-scoped). Only one should land — this one and #110863 overlap almost fully. Please close/dedupe before merging.
- `attach()` re-raises after detach while #110861 returns `False`: confirm the `pty_ws` caller treats the raised path identically (close-and-detach branch) so the exception doesn't surface as a 500/noise log. Non-blocking if verified.

**Verdict:** Looks good to merge, subject to the dedupe decision with #110861/#110863.

*Reviewed using Hermes-Agent*

**Comment by MohamadKanso** [comment]
thanks, i fixed the replay path based on your note. attach now detaches the failed socket and returns false, so pty_ws uses its existing close-and-detach path instead of leaking the exception. live send failures still use the identity check, so a stale socket cannot detach a replacement. the pty and reconnect tests are green: 20 passed. pushed in 3c3764a768.

**Comment by MohamadKanso** [comment]
closing this original PR in favor of merged #111357, which carried both implementation commits forward with the dashboard follow-ups.
